text/0034-ternary-hash.md

```diff
@@ -0,0 +1,415 @@
++ Feature name: `ternary-hash`
++ Start date: 2019-10-15
++ RFC PR: [iotaledger/bee-rfcs#34](https://github.com/iotaledger/bee-rfcs/pull/34)
++ Bee issue: [iotaledger/bee#77](https://github.com/iotaledger/bee/issues/77)
+
+# Summary
+
+This RFC proposes the ternary `Hash` type, the `Sponge` trait, and two cryptographic hash functions `CurlP` and `Kerl`
+implementing it. The 3 cryptographic hash functions used in the IOTA current networks (i.e. as of IOTA Reference
+Implementation `iri v1.8.1`) are `Kerl`, `CurlP27`, and `CurlP81`.
+
+Useful links:
+
++ [The sponge and duplex constructions](https://keccak.team/sponge_duplex.html)
++ [Curl-p](https://github.com/iotaledger/iri/blob/dev/src/main/java/com/iota/iri/crypto/Curl.java)
++ [Kerl specification](https://github.com/iotaledger/kerl/blob/master/IOTA-Kerl-spec.md)
++ [`iri v1.8.1`](https://github.com/iotaledger/iri/releases/tag/v1.8.1-RELEASE)
++ [IOTA transactions and bundle](https://docs.iota.org/docs/dev-essentials/0.1/concepts/bundles-and-transactions)
+
+# Motivation
+
+In order to participate in the IOTA network, an application needs be able to construct valid messages that can be
+verified by nodes in the network. Conversely, a node needs to be able to verify other nodes messages. Among other
+operations, this is accomplished by checking transaction signatures (from which addresses are calculated), transaction
+hashes, and hashes of collections of transactions (so-called bundles).
+
+The two hash functions currently used are both sponge constructions: `CurlP`, which is specified entirely in balanced
+ternary, and `Kerl`, which first converts ternary input to a binary representation, applies `keccak-384` to it, and then
+converts its binary output back to ternary. For `CurlP` specifically, its variants `CurlP27` and `CurlP81` are used.
+
+# Detailed design
+
+## Hash
+
+This RFC defines a ternary type `Hash`. The exact definition is implementation detail but an example definition could
+simply be the following.
+
+```rust
+struct Hash([i8; 243]);
```

Comment:
Why is Hash defined as `[i8; 243]` and transaction fields like Address like this: `Address(pub(crate)** TritBuf<T1B1Buf>)`?

Reply:
I think for the sake of this RFC we don't really care since this is implementation detail.

But for the actual answer I invoke @zesterer